src/BTree.mo

```diff
@@ -0,0 +1,139 @@
+
+import A "Array";
+import I "Iter";
+import List "List";
+import Option "Option";
+import Order "Order";
+import P "Prelude";
+import Debug "Debug";
+import Prim "mo:⛔";
+
+module {
+
+  /// Constants we use to shape the tree.
+  /// See https://en.wikipedia.org/wiki/B-tree#Definition
+  module Constants {
+    let MAX_CHILDREN = 4;
+  };
+
+  public type Compare<K> = {
+    show : K -> Text;
+    compare : (K, K) -> Order.Order
+  };
+
+  public type Data<K, V> = [(K, V)];
+
+  public type Internal<K, V> = {
+    data : Data<K, V>;
+    trees : [Tree<K, V>];
+  };
+
+  public type Tree<K, V> = {
+    #internal : Internal<K, V>;
+    #leaf : Data<K, V>;
+  };
+
+  func find_data<K, V>(data : Data<K, V>, find_k : K, c : (K, K) -> Order.Order) : ?V {
+    for ((k, v) in data.vals()) {
+      if (c(k, find_k) == #equal) { return ?v };
+    };
+    return null
+  };
+
+  public func find<K, V>(t : Tree<K, V>, k : K, c : (K, K) -> Order.Order) : ?V {
+   switch t {
+      case (#leaf(d)) { return find_data<K, V>(d, k, c) };
+      case (#internal(i)) {
+        for (j in I.range(0, i.data.size())) {
+          switch (c(k, i.data[j].0)) {
+            case (#equal) { return ?i.data[j].1 };
+            case (#less) { return find<K, V>(i.trees[j], k, c) };
+            case _ { }
+          }
+        };
+        find<K, V>(i.trees[i.data.size()], k, c)
+      };
+    };
+  };
+
+  /// Check that a B-Tree instance observes invariants of B-Trees.
+  /// Invariants ensure performance is what we expect.
+  /// For testing and debugging.
+  ///
+  /// Future refactoring --- Eventually, we can return Result or
+  /// Option so that both valid and invalid inputs can be inspected in
+  /// test cases.  Doing assertions directly here is easier, for now.
+  public module Check {
+
+    type Inf<K> = {#infmax; #infmin; #finite : K};
+
+    type InfCompare<K> = {
+      compare : (Inf<K>, Inf<K>) -> Order.Order;
+      show : Inf<K> -> Text
+    };
+
+    func infCompare<K>(c : Compare<K>) : InfCompare<K> = {
+      show = func (k : Inf<K>) : Text {
+        switch k {
+          case (#infmax) "#infmax";
+          case (#infmin) "#infmin";
+          case (#finite k) "#finite(" # c.show k # ")";
+        }
+      };
+      compare = func (k1 : Inf<K>, k2 : Inf<K>) : Order.Order {
+        switch (k1, k2) {
+        case (#infmin, _) #less;
+        case (_, #infmin) { /* nonsense case. */ assert false; loop { } };
+        case (_, #infmax) #less;
+        case (#infmax, _) { /* nonsense case. */ assert false; loop { } };
+        case (#finite(k1), #finite(k2)) c.compare(k1, k2);
+        }
+      }
+    };
+
+    public func root<K, V>(compare : Compare<K>, t : Tree<K, V>) {
+      switch t {
+        case (#leaf _) { rec(#infmin, infCompare(compare), t, #infmax) };
+        case (#internal i) {
+          if (i.data.size() == 0) { assert i.trees.size() == 0; return };
+          if (i.trees.size() < 2) { assert false };
```

Comment:
Is there a benefit to having all of the invalid cases trap vs. have be this be more like a `isBTreeValid()` function?

I guess the question from a usability standpoint is do you see developers using it like

```
try {
  Check.root<Text, Text>(Text.compare, bt);
  bt;
catch e {
  // invalid btree logic here
}
```

or 

```
do {
  if (isBTreeValid<Text, Text>(Text.compare, bt) {
    bt
  } else {
    // invalid btree logic here
  }
}

Reply:
Using `assert` is easier for this phase of development.

Indeed, using `assert` does not lead to composable code (the second version of the code you gave wouldn't work), so eventually, to make these checks more composable and thus more useful, they should either return `? Error` or perhaps `Result<_, Error>` for some well-defined variant type of `Error` (to be defined) -- not sure about the value of returning anything in the `#ok` case, so that's why I'd tend toward returning `? Error` and have `null` mean "okay", no errors.  Then, a wrapper can check for `null` and return `true`, or some other kind of `Result`, etc.